### ISSUE TITLE
Is Cython for performance a 'good idea'?

### ISSUE THREAD

**Issue by kodonnell** [PyCQA/pylint#2405]
Is using Cython to enhance the performance of pylint (and astroid) a 'good idea'? Specifically:

- Would the maintainers simply not be interested? Is Cython unpopular because e.g. it makes it harder to distribute? Or harder to maintain and contribute to?
- Is it 'likely' to result in significant speedup? From my relatively minimal experience, I'd expect so - probably an order of magnitude without going too crazy. However, there may be some things that make it hard to Cythonize fully (e.g. `yields`)?
- Are there other 'better' ways to get a similar speedup? (E.g. caching, etc.)

The idea would be we're in a happy medium where some core performance critical code is Cython (with maybe a bit of C), while the bulk of the 'business logic' for different linting rules etc. are still in python (as these are the parts most likely to be contributed by the community etc.?). Then again, some of these tried-and-true ones could be cythonized too, etc.

**Comment by PCManticore** [comment]
Also I'm closing this not necessarily because I'm opposed to the idea, but we don't have as much time as this effort requires.

**Comment by kodonnell** [comment]
Thanks @PCManticore, and I understand where you're coming from. It's good to have this documented for others (e.g. @nickdrozd who mentioned this possibility in his PRs, if I remember correctly).

Question: if somebody (maybe me or maybe someone who reads this) did a proof-of-concept cythonizing of astroid, and the results were promising (e.g. minimal code change but performance is doubled) would that change the situation much? I guess what I'm really asking is whether there's much point in me investigating this further. Another alternative (which may be cleaner) could be for someone to start a Cython fork of astroid and, if it's up-to-scratch, simply allow pylint to use either one. In other words "_PyCQA will maintain and support pylint with astroid. We allow for the use of <cython fork of astroid> in pylint, but it's at your own risk and not supported_" etc. That could also allow you to gauge how much interest there is, etc.

Re type annotation - that wouldn't have any impact for me (or most of the open source projects I see which don't seem to use it - though that may change).

**Comment by PCManticore** [comment]
@kodonnell yes, definitely. If you or anyone else will come up with a Cythonized astroid fork that has improved performance, we'll be definitely keen on integrating that into astroid itself. I'd love a faster pylint, I just cannot commit myself on working on this project.

**Comment by kodonnell** [comment]
Quick update @PCManticore - I've done a [quick hack](https://github.com/kodonnell/astroid/commit/dcc00c69c40691db8f827890afed7e00c38f2234) to just cythonize the plain python - note that very little code has changed. If you want to run it: `python setup.py build_ext --inplace`. 

Note that I couldn't 'naively' apply this to some of the files (seemingly inference.py and raw_building.py if I remember rightly) due to the dynamic assignment of methods not working out-of-box in cython - see [here](https://github.com/cython/cython/wiki/FAQ#how-do-i-implement-a-single-class-method-in-a-cython-module). Nonetheless, when running it on `pycodestyle.py` the runtime is now about 5.8s instead of 6.2s (both after a bit of warmup), which is pretty nice for 'free'. Also, given you say inferencing is the slow part, I should expect more gains once I get `inference.py` cythonized? (Which doesn't appear too hard to fix, but I need to go to bed.) NB: tests are failing, but I'm hoping that's just related to similar cython foibles. I don't think it was shortcutting the execution (hence faster timing) given it gave me the same score in both methods.

Also - do you want me to move this issue to astroid?

**Comment by kodonnell** [comment]
Great, I'll start a WIP PR or something this weekend.

It's worth noting that this same sort of naive cythonization could be applied to pylint - though my understanding is that astroid's the main bottleneck, so it might not be worth it.

**Comment by kodonnell** [comment]
@PCManticore do you think this should be re-opened given the results above?
